File: services/endpoints/worker/functions/relevancy-compute copy.py
import json
import argparse
import numpy as np
import chromadb
from sentence_transformers import SentenceTransformer

# Constants
MODEL_NAME = 'sentence-transformers/all-MiniLM-L6-v2'
COLLECTION_NAME = "paper-embeddings"
pathRefPapers = '/Users/spankyed/Develop/Projects/CurateGPT/services/files/assets/ref-papers.json'

# Load SBERT model
model = SentenceTransformer(MODEL_NAME)

def store_paper_embeddings_in_chroma(papers, collection_name=COLLECTION_NAME):
    client = chromadb.Client()
    
    # Check if the collection already exists, if not, create it
    if collection_name not in client.list_collections():
        client.create_collection(collection_name)

    embeddings = [model.encode(paper['title'] + ". " + paper['abstract']) for paper in papers]

    client.get_collection(collection_name).add(
        documents=embeddings,
        metadatas=papers,
        ids=[paper['id'] for paper in papers]
    )

def get_relevancy_scores(papers, n_results=5):
    client = chromadb.Client()
    # Assume if collection doesn't exist, embeddings need to be stored
    if COLLECTION_NAME not in chromadb.Client().list_collections():
        with open(pathRefPapers, 'r') as file:
            ref_papers = json.load(file)
        store_paper_embeddings_in_chroma(ref_papers)

    collection = client.get_collection(COLLECTION_NAME)

    for paper in papers:
        query_embedding = model.encode(paper['title'] + ". " + paper['abstract'])
        results = collection.query(query_texts=[query_embedding], n_results=n_results)
        
        relevancy_scores = [res['score'] for res in results[0]]
        avg_relevancy = sum(relevancy_scores) / len(relevancy_scores)

        # Add the relevancy score to the paper's metaData
        if 'metaData' not in paper:
            paper['metaData'] = {}
        paper['metaData']['relevancy'] = avg_relevancy

    return papers

File: services/endpoints/worker/functions/relevancy-compute copy.ts
// import json;
import { ChromaClient } from 'chromadb';
import { pipeline } from 'transformers.js';  // Assuming this is the correct import for transformers.js

// Constants
const MODEL_NAME = 'Xenova/all-MiniLM-L6-v2';
const COLLECTION_NAME = "paper-embeddings";
const PATH_REF_PAPERS = '/Users/spankyed/Develop/Projects/CurateGPT/services/files/assets/ref-papers.json';

const client = new ChromaClient();

class SBertEmbeddingFunction {
    private extractor: any;

    constructor(modelName: string) {
        this.extractor = pipeline('feature-extraction', modelName);
    }

    public async generate(texts: string[]): Promise<number[][]> {
        const embeddings: number[][] = await Promise.all(texts.map(async text => {
            const output = await this.extractor(text, { pooling: 'mean', normalize: true });
            return Array.from(output.data) as number[];
        }));
        return embeddings;
    }
}

const embedder = new SBertEmbeddingFunction(MODEL_NAME);

async function storePaperEmbeddingsInChroma(papers: any[], collectionName: string = COLLECTION_NAME) {
    const embeddings = await embedder.generate(papers.map(paper => paper.title + ". " + paper.abstract));
    
    if (!client.listCollections().includes(collectionName)) {
        client.createCollection(collectionName);
    }
    const collection = client.getCollection(collectionName);
    collection.add({
        documents: embeddings,
        metadatas: papers,
        ids: papers.map(paper => paper.id)
    });
}

async function getRelevancyScores(papers: any[], nResults = 5): Promise<any[]> {
    if (!client.listCollections().includes(COLLECTION_NAME)) {
        const refPapers = JSON.parse(fs.readFileSync(PATH_REF_PAPERS, 'utf-8'));
        await storePaperEmbeddingsInChroma(refPapers);
    }

    const collection = await client.getCollection(COLLECTION_NAME);
    for (let paper of papers) {
        const queryEmbedding = await embedder.generate([paper.title + ". " + paper.abstract]);
        const results = await collection.query({ queryTexts: queryEmbedding, nResults: nResults });
        
        const relevancyScores = results[0].map(res => res.score);
        const avgRelevancy = relevancyScores.reduce((a, b) => a + b, 0) / (relevancyScores.length || 1);

        paper.metaData = paper.metaData || {};
        paper.metaData.relevancy = avgRelevancy;
    }

    return papers;
}
